dask/dataframe/io/parquet.py

```diff
@@ -540,16 +582,174 @@ def _read_pyarrow(fs, fs_token, paths, columns=None, filters=None,
+    """
+    Convert an input time in the specified units to nanoseconds
+
+    Parameters
+    ----------
+    val: int
+        Input time value
+    unit : str
+        Time units of `val`.
+        One of 's', 'ms', 'us', 'ns'
+
+    Returns
+    -------
+    int
+        Time val in nanoseconds
+    """
+    if unit == 's':
+        factor = int(1e9)
+    elif unit == 'ms':
+        factor = int(1e6)
+    elif unit == 'us':
+        factor = int(1e3)
+    elif unit == 'ns':
+        factor = 1
+    else:
+        raise ValueError("Unsupported time unit '{unit}'"
+                         .format(unit=unit))
+
+    return val * factor
+
+
+def _get_pyarrow_divisions(pa_pieces, divisions_name, pa_schema, infer_divisions):
+    """
+    Compute DataFrame divisions from a list of pyarrow dataset pieces
+
+    Parameters
+    ----------
+    pa_pieces : list[pyarrow.parquet.ParquetDatasetPiece]
+        List of dataset pieces. Each piece corresponds to a single partition in the eventual dask DataFrame
+    divisions_name : str|None
+        The name of the column to compute divisions for
+    pa_schema : pyarrow.lib.Schema
+        The pyarrow schema for the dataset
+    infer_divisions : bool or None
+        If True divisions must be inferred (otherwise an exception is raised). If False or None divisions are not
+        inferred
+    Returns
+    -------
+    list
+    """
+    # Local imports
+    import pyarrow as pa
+    import pyarrow.parquet as pq
+
+    if infer_divisions is True and pa.__version__ < distutils.version.LooseVersion('0.9.0'):
+        raise NotImplementedError('infer_divisions=True requires pyarrow >=0.9.0')
+
+    # Check whether divisions_name is in the schema
+    # Note: get_field_index returns -1 if not found, but it does not accept None
+    if infer_divisions is True:
+        divisions_name_in_schema = divisions_name is not None and pa_schema.get_field_index(divisions_name) >= 0
+
+        if divisions_name_in_schema is False and infer_divisions is True:
+            raise ValueError(
+                'Unable to infer divisions for because no index column was discovered')
+    else:
+        divisions_name_in_schema = None
+
+    if pa_pieces and divisions_name_in_schema:
+        # We have pieces and a valid division column.
+        # Compute min/max for column in each row group
+        min_maxs = []
+        last_max = None
+
+        # Initialize index of divisions column within the row groups.
+        # To be computed during while processing the first piece below
+        divisions_col_index = None
+
+        for piece in pa_pieces:
+            pf = piece.get_metadata(pq.ParquetFile)
+            rg = pf.row_group(0)
+
+            # Compute division column index if needed
+            if divisions_col_index is None:
+                rg_paths = [rg.column(i).path_in_schema for i in range(rg.num_columns)]
+                try:
+                    divisions_col_index = rg_paths.index(divisions_name)
```

Comment:
Will this be the same for all of the pa_pieces?

Reply:
Good question. This assumption might be violated if someone passed in a list of paths to parquet datasets where some of the datasets are partitioned and others aren't.  Since I'm not confident, I switched back to computing this for each piece.